### ISSUE TITLE
[HOLD App#11795] [$500] Pressing enter takes the focus to Add a personal message field on Android - reported by @thesahindia

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#10181]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open :expensify: at mobile browser on Android
2. Go to settings > workspace > manage members > invite
3. Press enter on keyboard

Alternative Flow 
1. Open :expensify: at mobile browser
2. Go to settings > workspace > manage members > invite
3. Tap on Add a personal message field and press enter

## Expected Result:

- Pressing enter should not change the focus to the personal message field
- Pressing enter a first time should select the first contact
- Pressing enter another time should select the next contact in the list

https://user-images.githubusercontent.com/87341702/183491511-51828fc9-3c55-4372-a83f-1d63d4769502.mov

## Actual Result:
On pressing enter the  Add a personal message field gets focused

The user can't go to the next line and pressing enter selects the contacts

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Mobile Web (Android)
- Android

Make sure this works same on iOS too and matches desktop/browser behaviour.

**Version Number:** 1.1.85-8
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43995119/182292306-2b774b96-88d8-4259-8527-a1f246d59569.mp4

potentially related to https://expensify.slack.com/archives/C01GTK53T8Q/p1657397089778049 

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @thesahindia
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1657396762843809

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MitchExpensify** [comment]
I am unable to reproduce this on iOS

**Comment by thesahindia** [comment]
> I am unable to reproduce this on iOS

Only reproducible at android mobile browsers

**Comment by MitchExpensify** [comment]
Asked in Slack for help with reproduction - https://expensify.slack.com/archives/C01GTK53T8Q/p1659456631390199?thread_ts=1657396762.843809&cid=C01GTK53T8Q

**Comment by MitchExpensify** [comment]
@rushatgabhane  reproduced on v1.1.86-5, Chrome, Android 12

This is odd behavior, labelling Engineering accordingly

**Comment by Justicea83** [comment]
this can be external

**Comment by MitchExpensify** [comment]
[Upwork Job](https://www.upwork.com/jobs/~01ae3835d3fb9f6447)

**Comment by railway17** [comment]
### Proposal

I could replicate the issue in only real devices, not simulators.

Actually, this is an exceptional case, which we can reproduce even though we test with 2 input boxes in an empty page.
So, we can resolve this issue by adding `returnKeyType` props in the `OptionsSelector` like this.

```diff
const textInput = (
            <TextInput
                ref={el => this.textInput = el}
                value={this.props.value}
                label={this.props.textInputLabel}
                onChangeText={(text) => {
                    if (this.props.shouldFocusOnSelectRow) {
                        this.textInput.setNativeProps({selection: null});
                    }
                    this.props.onChangeText(text);
                }}
                placeholder={this.props.placeholderText || this.props.translate('optionsSelector.nameEmailOrPhoneNumber')}
                onBlur={(e) => {
                    if (!this.props.shouldFocusOnSelectRow) {
                        return;
                    }
                    this.relatedTarget = e.relatedTarget;
                }}
                selectTextOnFocus
+               returnKeyType="done"
                blurOnSubmit={Boolean(this.state.allOptions.length)}
            />
```

It will resolve this issue.

**Comment by aldo-expensify** [comment]
I was trying to test @railway17 's proposal, but the App for android is extremely slow at the moment (https://expensify.slack.com/archives/C01GTK53T8Q/p1659863626602549)

I updated the expected result in the GH issue to be more precise

**Comment by railway17** [comment]
> I was trying to test @railway17 's proposal, but the App for android is extremely slow at the moment (https://expensify.slack.com/archives/C01GTK53T8Q/p1659863626602549)
> 
> I updated the expected result in the GH issue to be more precise

I don't think my proposal makes the App be delayed.
Also, there are several places to use the `returnKeyType` props such as `PasswordPopover.js`, `AddSecondaryLoginPage.js` etc.
If those are not problems to use, this proposal should work.

**Comment by aldo-expensify** [comment]
@railway17 sorry for the confusion, I didn't mean that your proposed change had a performance impact. I was just saying that I haven't been able to test because the app at the moment is really slow for me, maybe because of some unrelated changes or maybe because I have something misconfigured in my env.

**Comment by aldo-expensify** [comment]
> I do not find it very friendly. This feels strange to me.

Agreed, I wasn't expecting that behaviour either, but I thought that the web functionality of selecting the next contact with enter was intended and we wanted to be consistent in mobile.

> Moreover, this issue talks about the mWeb platform and suggested functionality is for the web.

not exactly sure what you mean with this: "suggested functionality is for the web", do you say that because the video was taken from web?

> In real, the enter key on the virtual keyboard is supposed to take you to the next input(default functionality of the browser). Do we want to change that?

This makes sense to me, but I guess we would do the same for web and mobile, right?

**Comment by aldo-expensify** [comment]
Thanks @parasharrajat for looking into it. 

> Also, on web enter key press does not react to search input instead it selects the highlighted row(not necessarily next to the previous, the user can change it via arrow keys).

ah! I didn't notice, good to know

> I consider Enter key events to be suitable for web interface as the user's primary input is Keyboard but on mobile it is touch.
>
> I have never Used the keyboard enter key to select or deselect anything on the Mobile. Instead there Pressing on Enter makes more sense as the final submission. So that's why I think the behavior is different on Mobile from the web.

Hmm what you say makes sense to me, so.. one thing I was assuming is that we wanted the same behaviour for web/mobile, but this assumption may be wrong because there are cases where having a different behaviour can make sense, right? If that is correct, I guess my correction of the "Expected results" in the GH issue is wrong, and we would have to define how we want this to behave in mobile, a maybe the current behaviour is just fine.

**Comment by MitchExpensify** [comment]
>  I was assuming is that we wanted the same behaviour for web/mobile

+1 to this, we should have the same behavior on mobile and web in this case right? I cannot think of a reason we wouldn't. (So we'd want pressing enter to pick the highlighted user)

> I tested this and the app is behaving correctly as per the expected result. Pressing enter selects the highlighted user and moves to the next target same as native on mWeb Chrome.

Just to confirm @parasharrajat, are you saying you pressed enter on mobile and it highlighted the user top of the list? I just tried this on iOS and didn't experience that (It just stopped focusing on the search field)

**Comment by MitchExpensify** [comment]
Did you get a chance to retest this @parasharrajat? Tempted to close this out if not

**Comment by aldo-expensify** [comment]
hehe, I guess that is up to us to decide... at the start I was assuming that we wanted web (desktop and mobile) and mobile native to behave exactly the same, but [this comment](https://github.com/Expensify/App/issues/10181#issuecomment-1212418978) from @parasharrajat made me consider that this may not be the case always.

I don't feel I have enough experience with UI/UX to decide on what is the best user experience here. I'll ask @shawnborton , maybe he may have a stronger opinion on whats best.

**Comment by shawnborton** [comment]
I'm a little bit confused about what this bug is. It says that "Pressing enter should not change the focus to the personal message field" but when looking at the original video, it doesn't seem like the focus moves from the top text input to the personal message field. Can someone clarify what's happening here?

**Comment by shawnborton** [comment]
Got it, thank you! I would think hitting enter would keep you focused into that field - if you want to go to the next one, you would need to hit tab.

**Comment by shawnborton** [comment]
Fair point... I think we'd want to try to optimize for platform parity though, so in this case, I would think we'd follow web/desktop. Thoughts?

**Comment by MitchExpensify** [comment]
Did you get a chance to follow up on the analysis menioned above @parasharrajat ?

**Comment by railway17** [comment]
Do you think my proposal will not work? @parasharrajat

**Comment by railway17** [comment]
> As far as I can tell, the proposed solution will close the keyboard when the done key is pressed. Right?
> 
> if that is the case, no! that does not satisfy the expected result.

No, it will not close the keyboard

**Comment by railway17** [comment]
> Ok,let me check.

I have just tested again. It doesn't close the keyboard when clicking Enter

**Comment by mountiny** [comment]
I think that does not need ot be part of this issue. 

@parasharrajat @railway17 Since there was a lot of discussion, could you please once more recap what is the current behaviour and what would be the final behaviour with your change (sorry if it is clear for you, just wanted to get everyone on the same page before we head into implementation)?

**Comment by railway17** [comment]
Ok
According to my replication, the issue is displaying in real Android devices. (Invite New Members page)
There are 2 input boxes (`name, email, phone number` and `personal message` fields).
While inserting `name, email, phone number` value and clicking Enter, the focus will be moved to `personal message` field.
It is an **actual issue**.

![image](https://user-images.githubusercontent.com/29966461/190640882-6fe40605-cf43-4d03-9d95-719706549b2a.png)

[My proposal](https://github.com/expensify/App/issues/10181#issuecomment-1204954480) will resolve this issue.
That's all 🍔

**Comment by railway17** [comment]
> @railway17 Thanks, can you describe (in steps) what will happen if we press Enter while focused in the `name, email, phone number` field when your solution is implemented?

It keeps `name, email, phone number` field be focused.

**Comment by mountiny** [comment]
@railway17 Coming from the Expected results section, will it select the first contact in the list?

**Comment by railway17** [comment]
> @railway17 Coming from the Expected results section, will it select the first contact in the list?

Looks like `Expected Result` was changed after I sent my proposal 👀

**Comment by railway17** [comment]
> Ok, Thanks for clarification @mountiny. But this change does not work on Android. As I said earlier, it closes the keyboard that's all.

Can you show me your replication video?
In my side, it is still not closing the keyboard

**Comment by railway17** [comment]
@parasharrajat 
Thank you for your sharing video.
It's very strange. What is your Android version?

**Comment by railway17** [comment]
> If you see closly that video is for Android native but your proposal is working fine on Android web.

Ah..
I thought you were referring Android web.
Ok, I didn't test in Android native.
Thank you.

**Comment by mountiny** [comment]
> Looks like Expected Result was changed after I sent my proposal 👀

Ah sorry, thats is why I wanted to confirm, there was lots of discussion and often in these cases the Expected result section might not reflect what we want of what the proposal actually does.

**Comment by railway17** [comment]
Ok, let me try again.
For now, I got a new `**Expected Result**`.

**Comment by thesahindia** [comment]
The keyboard closes on pressing enter at Android for all the pages where we have contact list so I think the contact selection feature doesn't work at Android and if we wanna allow it we should do it for other places too? E.g. new chat page

**Comment by railway17** [comment]
> The keyboard closes on pressing enter at Android for all the pages where we have contact list so I think the contact selection feature doesn't work at Android and if we wanna allow it we should do it for other places too? E.g. new chat page

Yes, @parasharrajat 
The keyboard is closing in Native app even though even though don't use my change 👀

**Comment by railway17** [comment]
So, @thesahindia , @parasharrajat 
Do you think we should implement here as original `expected result`?
And need to implement a new `expected result` in another ticket?

**Comment by railway17** [comment]
Btw, don't you have speed issue while trying native app?
It's very hard to test with app in my side because speed is too slow.

**Comment by mountiny** [comment]
I have updated the title and the PR body to clarify we want their behaviour same on all the platforms.

@MitchExpensify would you be able to double price on this one? Thank you!

**Comment by MitchExpensify** [comment]
Doubled and had to create a new [Upwork Job](https://www.upwork.com/jobs/~0139b06154e407e923)

**Comment by railway17** [comment]
@mountiny 
It's strange but I've noticed that it is working on my side.
In both real and simulator devices.
Let me attach my video that was taken from the simulator.
The testing environment is: staging.new.expensify.com

https://user-images.githubusercontent.com/29966461/191411807-4be40a3a-b464-467c-b8fb-5335703b151f.mov

**Comment by mountiny** [comment]
@railway17 Thanks for raising this.

@mvtglobally Could your team please confirm this is working as shown in the [latest screenshare](https://github.com/Expensify/App/issues/10181#issuecomment-1253176444) in Android?

If yes, than we can close this issue out.

**Comment by mvtglobally** [comment]
@mountiny seem to be still happening. Not sure if we are missing anything

https://user-images.githubusercontent.com/43995119/192201718-66ca663f-ec97-456f-8a68-39a5cef0f07a.mp4

**Comment by MitchExpensify** [comment]
Not overdue

**Comment by railway17** [comment]
@laurenreidexpensify 
This issue doesn't need proposal anymore?
I've got solution and would like to update my proposal

**Comment by mountiny** [comment]
@railway17 Sorry, we have updated our process and the Exported label was removed, we will only use External label from now on. 

I think you proposal did not cover all of the bullets of the expected results right? We would like to unify the behaviour across al the platforms. cc @parasharrajat for confirmation, I am getting quite lost in this issue.

**Comment by railway17** [comment]
> @railway17 Sorry, we have updated our process and the Exported label was removed, we will only use External label from now on.
> 
> I think you proposal did not cover all of the bullets of the expected results right? We would like to unify the behaviour across al the platforms. cc @parasharrajat for confirmation, I am getting quite lost in this issue.

Yes, So, I find out new solution and to cover all of the bullets.
And would like to propose if this issue still requires external help. (It means hiring via upwork like other issues)

**Comment by mountiny** [comment]
It does @railway17, the External label and Help Wanted labels are labels to look out for.

Please feel free to share the proposal. Thank you!

**Comment by railway17** [comment]
### Proposal (Updated)

Here is an updated proposal based on the new requirements.
I could reproduce this issue in lower versions of Android.
Here is [video](https://drive.google.com/file/d/1psuo7UF83MFi9bpJPaN1d4MsmT2wFPbb/view?usp=sharing) that I tested. (_The left one is API version 26, and the right one is 30. In the current video, the right one has not latest chrome version, but I tested it again after installing the latest version on it. It was also working).
As you can see, this issue is reproducible in lower versions. 

mWeb issue will be fixed by ` returnKeyType="done"` as we discussed in above comments.

**For other platforms will resolve like below:**

Important is that we have to customize the action handler of `Enter` key.
I confirmed that this change will work in the same way on all the platforms such as Android, iPhone, Desktop, mWeb (Android Chrome and iPhone Safari).
So, When clicking the `Enter` key, it will be selected from ahead one by one. 

To do this:

1. In the `src/components/OptionsSelector/BaseOptionsSelector.js`, update the props of Search TextInput:
- Change the `blurOnSubmit` props to `false`
- Define method for `onSubmitEditting`
- Change the selected option list from ahead one by one when clicking the Enter key

Changed Code will be like below:


```diff
...
+    /** Toggle select the options from ahead one by one when clicking Enter key in mobile */
+    toggleOptionFromAhead: PropTypes.func,
...
    shouldDelayFocus: false,
+  toggleOptionFromAhead: ()=> {},
    ...optionsSelectorDefaultProps,

...
+    onSubmit = () => {
+        this.props.toggleOptionFromAhead()
+        this.setState({
+            focusedIndex: this.props.selectedOptions.length,
+        });
+   }
...
                     this.relatedTarget = e.relatedTarget;
                 }}
                 selectTextOnFocus
-                blurOnSubmit={Boolean(this.state.allOptions.length)}
+                returnKeyType="done"
+                blurOnSubmit={false}
+                onSubmitEditing={this.onSubmit}
             />
         );
...
```

2. In the `src/pages/workspace/WorkspaceInvitePage.js`, define new method to update the selected list.

```diff
...
        this.getExcludedUsers = this.getExcludedUsers.bind(this);
        this.toggleOption = this.toggleOption.bind(this);
+      this.toggleOptionFromAhead = this.toggleOptionFromAhead.bind(this)
        const {
            personalDetails,
            userToInvite,

...
+    /**
+    * 
+     * @param {number} index
+     */
+     toggleOptionFromAhead() {
+        this.clearErrors();
+
+        const filterText = _.reduce(this.state.selectedOptions, (str, {login}) => `${str} ${login}`, '');
+        const personalDetailsWithoutSelected = _.filter(this.state.personalDetails, ({login}) => !filterText.includes(login));
+        
+        if (personalDetailsWithoutSelected.length) {
+            this.setState((prevState) => {
+                 let newSelectedOptions = [...prevState.selectedOptions, personalDetailsWithoutSelected[0]];
+        
+                const {
+                    personalDetails,
+                    userToInvite,
+                } = OptionsListUtils.getMemberInviteOptions(
+                    this.props.personalDetails,
+                    this.props.betas,
+                    prevState.searchValue,
+                    this.getExcludedUsers(),
+                );
+    
+                return {
+                    selectedOptions: newSelectedOptions,
+                    personalDetails,
+                    userToInvite,
+                    searchValue: prevState.searchValue,
+                };
+            });
+        }
+    }

...
                                    headerMessage={headerMessage}
+                                   toggleOptionFromAhead={this.toggleOptionFromAhead}
                                    hideSectionHeaders
                                    hideAdditionalOptionStates


```

**Comment by parasharrajat** [comment]
Aha, it seems all settled, I don't have to answer the old question. 

Thanks for the proposal. I will review it. This is a little involved.

**Comment by mountiny** [comment]
@parasharrajat any updates?

**Comment by parasharrajat** [comment]
Monday is just started for me. I will review it asap.

**Comment by parasharrajat** [comment]
Still busy with other stuff. I will keep you posted here.

**Comment by mountiny** [comment]
any updates @parasharrajat?

**Comment by parasharrajat** [comment]
Will review this today?

**Comment by parasharrajat** [comment]
@railway17's proposal can be improved. it is missing proper explanation and not easy to understand. (Please try to make it easier for a reviewer to understand your proposal). The easier it is to understand, the better odds for you. 

Could you please add the following details into it? 
1. What behaviour are proposing? 
2. Explain how you will acheive it in in steps and then use code to support your explanation. 
3. There are different secenarios involved in that page. Explain, how is your proposal supporting those? 
 for example, if the first option in the list is already selected, what will happen when pressing enter? etc. 
 
When do you mean by `selected from ahead one by one`?

**Comment by mountiny** [comment]
@parasharrajat @shawnborton @MitchExpensify @aldo-expensify I am unsure if we should just proceed with this one on its own. Seems like the solution to it is quite involved and we have other similar forms/lists in the App which should have the same behaviour and I dont think we should only focus on this one solely.

The solution/behaviour should be same for all instances of this input form. At the same time, this feels like low ROI at the moment with out attention to better performance of the app.

I think it would be better to put this on HOLD and have some predesign about how this input form should behave across the platforms, clearly define all the scenarios and how they should be handled (similarly as we are doing with the navigation now), otherwise we will have many of similar issues like this where we will be trying to tweak the behaviour slightly depending on who the CME is 😄 Thoughts?

**Comment by shawnborton** [comment]
I like the idea of putting this on HOLD while we design a better solution.

**Comment by MitchExpensify** [comment]
Would leading that pre-design typically fall on the CM or CME? This is the first time I've run into a case like this as CM

**Comment by shawnborton** [comment]
I think we'd want to put this back on our internal design team to come up with a better solution, and probably just close this one out altogether?

**Comment by aldo-expensify** [comment]
> I think we'd want to put this back on our internal design team to come up with a better solution, and probably just close this one out altogether?

Sounds good to me!

**Comment by MitchExpensify** [comment]
> I think we'd want to put this back on our internal design team to come up with a better solution, and probably just close this one out altogether?

Cool! I'm tempted to pay out half price ($250) to C+ and the reporter of the issue for their work on this. What do you reckon @michaelhaxhiu ? (CM buddy check)

**Comment by mountiny** [comment]
> Cool! I'm tempted to pay out half price ($250) to C+ and the reporter of the issue for their work on this. What do you reckon @michaelhaxhiu ? (CM buddy check)

I agree with paying that out and I dont think either of us necessarily needs to lead the predesign we should just make sure this does not fall through the cracks and it gets done!

**Comment by mountiny** [comment]
Created a planning issue for this here https://github.com/Expensify/App/issues/11795 so we dont forget about it. Feel free to subscribe to it or take it on if you want to lead this sooner than later!

**Comment by michaelhaxhiu** [comment]
Let's take the pricing question to the internal #contributor-plus channel and discuss there. It would be good to keep track of precedent here in case similar situations arise in the future.

**Comment by MitchExpensify** [comment]
> @MitchExpensify Let's take the pricing question to the internal #contributor-plus channel and discuss there. It would be good to keep track of precedent here in case similar situations arise in the future.

[Asked here](https://expensify.slack.com/archives/C02NK2DQWUX/p1665770562144299)

**Comment by mountiny** [comment]
Put this on hold as mentioned before

**Comment by tjferriss** [comment]
Following instructions for the [weekly update chore](https://stackoverflow.com/c/expensify/questions/8208):

@mountiny this is one of the oldest issues in the /App repo. To help us clear out the large backlog of bugs, can you:

- Decide whether any proposals currently meet our guidelines and can be approved as-is
- For any that can't, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #bug-zero

I recognize this issue is on hold so maybe the above is irrelevant, but once it's not on hold do you expect we should take this internal given the issue's age?

**Comment by mountiny** [comment]
Thanks for the ping, in this particular case I will go ahead and close the issue. My reasons are:

- this is not strictly bug, but it can be considered as feature request.
- this is not blocking users from using the app
- this influences only Android
- we will fix this and similar issues related to option list selection in this issue https://github.com/Expensify/App/issues/11795

**Comment by parasharrajat** [comment]
@MitchExpensify Can you please look into the C+ payment https://github.com/Expensify/App/issues/10181#issuecomment-127684135

**Comment by MitchExpensify** [comment]
Hey @parasharrajat, apologies for the delay (Thanksgiving break then I was unwell) 

Invite sent to [new upwork job](https://www.upwork.com/jobs/~010e554dbc4be2e6d1)

**Comment by parasharrajat** [comment]
@MitchExpensify Could you please pay this off?
